src/protocol/modulus_conversion/convert_shares.rs

```diff
@@ -0,0 +1,180 @@
+use crate::{
+    error::BoxError,
+    field::Field,
+    helpers::{fabric::Network, prss::SpaceIndex},
+    protocol::{
+        context::ProtocolContext,
+        modulus_conversion::gen_random::{GenRandom, ReplicatedBinary},
+        reveal_additive_binary::RevealAdditiveBinary,
+        RecordId, Step,
+    },
+    secret_sharing::Replicated,
+};
+use futures::future::try_join_all;
+
+pub struct XorShares {
+    num_bits: u8,
+    packed_bits: u64,
+}
+
+pub struct ConvertShares {
+    input: XorShares,
+}
+
+///
+/// This is an implementation of
+/// Protocol 5.2 Modulus-conversion protocol from `Z_2^u` to `Z_p`
+/// from the paper <https://eprint.iacr.org/2018/387.pdf>
+///
+/// It works by generating two secret-sharings of a random number `r`,
+/// one in `Z_2`, the other in `Z_p`. The sharing in `Z_2` is subtracted
+/// from the input and the result is revealed.
+///
+/// If the revealed result is `0`, that indicates that `r` had the same value
+/// as the secret input, so the sharing in `Z_p` is returned.
+/// If the revealed result is a `1`, that indicate that `r` was different than
+/// the secret nput, so the sharing 1 - the sharing in `Z_p` is returned.
+impl ConvertShares {
+    #[allow(dead_code)]
+    pub fn new(input: XorShares) -> Self {
+        Self { input }
+    }
+
+    #[allow(dead_code)]
+    pub async fn execute<F: Field, S: Step + SpaceIndex, N: Network<S>>(
+        &self,
+        ctx: &ProtocolContext<'_, S, N>,
+        record_id: RecordId,
+        step0: S,
+        step1: S,
+        step2: S,
+        step3: S,
+    ) -> Result<Vec<Replicated<F>>, BoxError> {
+        let prss = &ctx.participant[step0];
+        let (left, right) = prss.generate_values(record_id.into());
+
+        let futures = (0..self.input.num_bits).into_iter().map(|i| async move {
+            let inner_record_id = RecordId::from(
+                u32::from(record_id) * u32::from(self.input.num_bits) + u32::from(i),
+            );
+            let b0 = left & (1 << i) != 0;
+            let b1 = right & (1 << i) != 0;
+
+            let input = self.input.packed_bits & (1 << i) != 0;
+
+            let input_xor_r = input ^ b0;
+
+            let r_binary = ReplicatedBinary::new(b0, b1);
+
+            let r_big_field: Replicated<F> = GenRandom::new(r_binary)
+                .execute(ctx, inner_record_id, step1, step2)
+                .await?;
+
+            let revealed_output =
```

Comment:
can we run this in parallel with `GenRandom`?

Reply:
Yes. I'll make it parallel for now, but long-term we should contemplate pre-generating them at non-peak times.